api/sessionrecording/session_recording.go

```diff
@@ -0,0 +1,316 @@
+/*
+Copyright 2024 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package sessionrecording can be used to read Teleport's
+// session recording files.
+package sessionrecording
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/binary"
+	"errors"
+	"io"
+	"log/slog"
+
+	"github.com/gravitational/trace"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+)
+
+const (
+	// Int32Size is a constant for 32 bit integer byte size
+	Int32Size = 4
+
+	// Int64Size is a constant for 64 bit integer byte size
+	Int64Size = 8
+
+	// MaxProtoMessageSizeBytes is maximum protobuf marshaled message size
+	MaxProtoMessageSizeBytes = 64 * 1024
+
+	// ProtoStreamV1 is a version of the binary protocol
+	ProtoStreamV1 = 1
+)
+
+const (
+	// maxIterationLimit is max iteration limit
+	maxIterationLimit = 1000
+)
+
+// NewReader returns a new session recording reader
+//
+// It is the caller's responsibility to call Close on the [Reader] when done.
+func NewReader(r io.Reader) *Reader {
+	return &Reader{
+		reader:    r,
+		lastIndex: -1,
+	}
+}
+
+const (
+	// protoReaderStateInit is ready to start reading the next part
+	protoReaderStateInit = iota
+	// protoReaderStateCurrent will read the data from the current part
+	protoReaderStateCurrent
+	// protoReaderStateEOF indicates that reader has completed reading
+	// all parts
+	protoReaderStateEOF
+	// protoReaderStateError indicates that reader has reached internal
+	// error and should close
+	protoReaderStateError
+)
+
+// Reader reads Teleport's session recordings
+type Reader struct {
+	inner      io.Reader
```

Comment:
choose a better name than inner. This holds the limit reader, which is only needed across read iterations to help catch dangling padding bytes.